libraries/AP_Scheduler/AP_Scheduler.tpp

```diff
@@ -0,0 +1,213 @@
+{
+    _tasks = nullptr;
+    _last_run = nullptr;
+  
+    AP_Param::setup_object_defaults(this, var_info);
+    _loop_rate_hz.set_and_save(loop_rate);
+
+    // only allow 50 to 400 Hz
+    if (_loop_rate_hz < 50) {
+        _loop_rate_hz.set(50);
+    } else if (_loop_rate_hz > 400) {
+        _loop_rate_hz.set(400);
+    }
+}
+
+// initialise the scheduler
+template <class Vehicle>
+AP_Scheduler<Vehicle>::~AP_Scheduler() {
+    if(_last_run) {
+        delete [] _last_run;
+    }
+    if(_perf_counters) {
+        delete [] _perf_counters;
+    }
+}
+
+// initialise the scheduler
+template <class Vehicle>
+void AP_Scheduler<Vehicle>::init(const AP_Task<Vehicle> *tasks, uint8_t num_tasks)
+{
+    if(tasks == nullptr || num_tasks == 0) {
+        return;
+    }
+    _tasks = tasks;
+    _num_tasks = num_tasks;
+    _last_run = new uint16_t[_num_tasks];
+    memset(_last_run, 0, sizeof(_last_run[0]) * _num_tasks);
+    _tick_counter = 0;
+}
+
+// one tick has passed
+template <class Vehicle>
+void AP_Scheduler<Vehicle>::tick(void)
+{
+    _tick_counter++;
+}
+
+/*
+  run one tick
+  this will run as many scheduler tasks as we can in the specified time
+ */
+template <class Vehicle>
+void AP_Scheduler<Vehicle>::run(uint32_t time_available)
+{
+    // break condition
+    if(_tasks == nullptr) {
+        return;
+    }
+  
+    if (_debug > 3 && _perf_counters == nullptr) {
+        _perf_counters = new AP_HAL::Util::perf_counter_t[_num_tasks];
+        for (uint8_t i=0; i<_num_tasks && _perf_counters != nullptr; i++) {
+            _perf_counters[i] = hal.util->perf_alloc(AP_HAL::Util::PC_ELAPSED, _tasks[i]._name);
+        }
+    }
+
+    for (uint8_t i=0; i<_num_tasks && _tasks[i]._max_time_micros <= time_available; i++) {
+        const uint16_t dt = _tick_counter - _last_run[i];
+        const uint16_t interval_ticks = (_loop_rate_hz / _tasks[i]._rate_hz) < 1 ? 1 : (_loop_rate_hz / _tasks[i]._rate_hz);
+                
+        if(dt < interval_ticks) {
+            continue;
+        }
+    
+        // run it
+        _task_time_started = AP_HAL::micros();
+        current_task = i;
+        if (_debug > 3 && _perf_counters[i]) {
+            hal.util->perf_begin(_perf_counters[i]);
+        }
+
+        // call that function :)
+        if(_parent) {
+            task_fn_t func = _tasks[i]._function;
+            (_parent->*func)();
+        }
+        
+        if (_debug > 3 && _perf_counters[i]) {
+            hal.util->perf_end(_perf_counters[i]);
+        }
+        current_task = -1;
+        
+        // record the tick counter when we ran. This drives
+        // when we next run the event
+        _last_run[i] = _tick_counter;
+
+        // work out how long the event actually took
+        const uint32_t time_taken = AP_HAL::micros() - _task_time_started;
+
+        // we've slipped a whole run of this task!
+        if (dt >= interval_ticks*2 && _debug > 1) {
```

Comment:
besides needing to be `_debug > 4` (which I think can actually remain as > 1), this misses the purpose of the check. This needs to be checked **before** running the task, like it is now.